src/index.ts

```diff
@@ -276,43 +282,68 @@ export type AppAnalysisResult = {
     mitmproxyEvents: MitmproxyEvent[];
 };
 
+/** Options for the emulator if you want it to be automatically managed by this library. */
+type StartEmulatorOptions = {
+    /** Whether to start the emulator in headless mode (default: `false`). */
+    headless?: boolean;
+    /** Whether to start the emulator with audio (default: `false`). */
+    audio?: boolean;
+    /** Whether to discard all changes when exiting the emulator (default: `true`). */
+    ephemeral?: boolean;
+    /** Options for hardware accelerations. These do not need to be set if everything works with the defaults. */
+    hardwareAcceleration?: {
+        /** Sets the `-accel` option, see https://developer.android.com/studio/run/emulator-commandline#common. */
+        mode?: 'auto' | 'off' | 'on';
+        /** Sets the `-gpu` option, see https://developer.android.com/studio/run/emulator-acceleration#accel-graphics. */
+        gpuMode?: 'auto' | 'host' | 'swiftshader_indirect' | 'angle_indirect' | 'guest';
+    };
+    /**
+     * How many times to retry starting the emulator in case of a failure.
+     *
+     * @default 1
+     */
+    attemptRestarts?: number;
+};
+
+export type AndroidEmulatorRunTargetOptions = (
```

Comment:
I'm assuming your comment about being open to breaking changes relates to this type? Do you have suggestions? I'm definitely not exactly happy with the current design.

Semi-related suggestion: How about instead of `createEmulator`, we use "managed" and "unmanaged emulator"?

Reply:
There are already braking changes because I moved `emulatorName` out of `StartEmulatorOptions`. But I have no better design in might, could you provide an example?

Reply:
Here's my go at a design. Unless I missed something it should even be fully backwards-compatible.

```ts
/** Options to configure how the emulator should be started by cyanoacrylate with the `emulator` command. */
export type StartEmulatorOptions = {
    /** Whether to start the emulator in headless mode (default: `false`). */
    headless?: boolean;
    /** Whether to start the emulator with audio (default: `false`). */
    audio?: boolean;
    /** Whether to discard all changes when exiting the emulator (default: `true`). */
    ephemeral?: boolean;
    /** Options for hardware accelerations. These do not need to be set if everything works with the defaults. */
    hardwareAcceleration?: {
        /** Sets the `-accel` option, see https://developer.android.com/studio/run/emulator-commandline#common. */
        mode?: 'auto' | 'off' | 'on';
        /** Sets the `-gpu` option, see https://developer.android.com/studio/run/emulator-acceleration#accel-graphics. */
        gpuMode?: 'auto' | 'host' | 'swiftshader_indirect' | 'angle_indirect' | 'guest';
    };
    /**
     * How many times to retry starting the emulator in case of a failure.
     *
     * @default 1
     */
    attemptRestarts?: number;
};

/** Run target options for an Android emulator that the user creates, manages, and starts themselves. */
export type AndroidEmulatorRunTargetOptionsUnmanagedSelfStarted = {
    /** The name of a snapshot to use when resetting the emulator. */
    snapshotName?: string;
};
/**
 * Run target options for an Android emulator that the user creates and manages themselves but that is automatically
 * started by cyanoacrylate.
 */
export type AndroidEmulatorRunTargetOptionsUnmanagedStarted = {
    /** Options to configure how the emulator should be started by cyanoacrylate. */
    startEmulatorOptions: {
        /** The name of the emulator to start. */
        emulatorName: string;
    } & StartEmulatorOptions;

    /** The name of a snapshot to use when resetting the emulator. */
    snapshotName?: string;
};
/** Run target options for an Android emulator that is created, managed, and started automatically by cyanoacrylate. */
export type AndroidEmulatorRunTargetOptionsManaged = {
    /** Options to configure how the emulator should be created by cyanoacrylate. */
    managedEmulatorOptions: {
        /**
         * A key to distinguish the emulator from other ones created by cyanoacrylate. All analyses using the same key
         * will share an emulator. The created emulator will be named `cyanoacrylate-{key}-{MD5 hash of the options}`.
         */
        key: string;

        /** Which kind of emulator to create. */
        type: EmulatorOptions;

        /** Optional honey data to place on the device before creating the snapshot. */
        honeyData?: {
            /** The device name to set, which shows up to other network or bluetooth devices. */
            deviceName?: string;
            /** Content to put into the clipboard. */
            clipboard?: string;
            /** Calendar events to place on the device. */
            calendarEvents?: CalendarEventData[];
            /** Contacts to place on the device. */
            contacts?: ContactData[];
        };

        /**
         * How often to try rebuilding the emulator completely in case of an error.
         *
         * @default 1
         */
        attemptRebuilds?: number;
    };

    /** Options to configure how the emulator should be started by cyanoacrylate. */
    startEmulatorOptions: StartEmulatorOptions;
};

/**
 * Run target options for an Android emulator. You can choose between the following variants:
 *
 * - The emulator is completely created, managed, and started automatically by cyanoacrylate. You declaratively specify
 *   the parameters of the emulator and cyanoacrylate ensures that an emulator matching those parameters is available
 *   after you run `analysis.ensureDevice()`.
 *
 *   It also automatically creates a snapshot of the emulator in a clean state for you that you can revert to using
 *   `analysis.resetDevice()`. You can optionally specify honey data that should be placed on the emulator before the
 *   snapshot is created.
 *
 *   Use {@link AndroidEmulatorRunTargetOptionsManaged}.
 * - You create and manage the emulator as well as the snapshot (if desired) yourself but cyanoacrylate automatically
 *   starts it for you. It also takes care to restart the emulator if necessary whenever you run
 *   `analysis.ensureDevice()`.
 *
 *   Use {@link AndroidEmulatorRunTargetOptionsUnmanagedStarted}.
 * - You create, manage, and start the emulator as well as the snapshot (if desired) yourself and cyanoacrylate does not
 *   touch it. You have to ensure yourself that the emulator is working and running. Use
 *   {@link AndroidEmulatorRunTargetOptionsUnmanagedSelfStarted}.
 */
export type AndroidEmulatorRunTargetOptions =
    | AndroidEmulatorRunTargetOptionsUnmanagedSelfStarted
    | AndroidEmulatorRunTargetOptionsUnmanagedStarted
    | AndroidEmulatorRunTargetOptionsManaged;
```

Reply:
Actually, we should probably default `managedEmulatorOptions.type` to Android 11, Google APIs, x86_64 since that is almost definitely the one people will want to use (https://github.com/tweaselORG/appstraction/issues/54).

Reply:
I pushed my suggestion on Monday. Yours looks really complicated. I am also not sure I want the `emulatorName`in the `startEmulatorOptions`, even though this would preserve the current API. It just seems out of place there.

Reply:
But the honey data is also not just a thing of emulators managed by ca, is it?

Reply:
> I pushed my suggestion on Monday. Yours looks really complicated.

This was in response to/based on your changes from Monday. You'll have to explain to me how this is more complicated other than being longer because of the docstrings. Imo this is much clearer (and will also produce much more readable output from typedoc).

> But the honey data is also not just a thing of emulators managed by ca, is it?

But in those cases, this declarative API doesn't make sense, does it? There are no snapshots that we could prepare there. I think it's fair that the user has to place the honey data themselves using the APIs provided by appstraction in those cases.